### PR TITLE
[span] Remove tuple interface

### PR DESCRIPTION
**Pull request by miscco**
# Description
This implements P2116R0 by removing the tuple interface from span.

Addresses #556

# Checklist

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [ ] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [x] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

### CODE REVIEW

**Comment by CaseyCarter** [comment]
@miscco, FYI "Addresses" has "link this PR from the issue" magic, but not "tell GitHub to close the issue when the PR is merged" magic. Had you instead used "Fixes" or "Resolves" I wouldn't have had to manually link the issue.

**Comment by miscco** [comment]
@CaseyCarter sorry for that. In my old job addresses was closing issues ;)

To keep track of it there is also a revision for [libcxx](https://reviews.llvm.org/D75640)

File: tests/std/tests/P0122R7_span/test.cpp [hunk 1]
```diff
@@ -966,10 +957,6 @@ constexpr bool test() {
         static_assert(is_same_v<decltype(sp_nine.rbegin()), span<int, 9>::reverse_iterator>);
         static_assert(is_same_v<decltype(sp_dyn.rend()), span<int>::reverse_iterator>);
         static_assert(is_same_v<decltype(sp_nine.rend()), span<int, 9>::reverse_iterator>);
-
-        static_assert(noexcept(get<5>(sp_nine)));
-        assert(get<5>(sp_nine) == 60);
-        assert(&get<5>(sp_nine) == begin(sequence) + 5);
     }
 
     {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0122R7_span/test.cpp]: I believe that the `auto& [x, y, z] = sp_three;` structured binding test also needs to be removed.

Comment (StephanTLavavej) [reply tests/std/tests/P0122R7_span/test.cpp]: Yep, verified it was failing to compile.

**Comment by miscco** [comment]
@StephanTLavavej Thanks for fixing the structured bindings.
I stupidely did not rebuild the whole STL when I compiled the tests.

**Comment by StephanTLavavej** [comment]
No problem, that's what I'm here for! :-)

Here's an STL maintainer trick you might not have thought of. When making changes to product code, of course it's best to rebuild the STL and rerun all of the tests. However, when you're making changes to header-only code which you expect to have limited impact, you can take a shortcut to iterate very rapidly. Simply compile a single test directly, and add `/I S:\GitHub\STL\stl\inc` (adjusting the repo path appropriately). This will "overlay" the repo's headers (which you're directly editing) over what's produced by the build system.

**Comment by StephanTLavavej** [comment]
I've submitted a Microsoft-internal PR for this; please don't push any more changes to this branch, and we should be able to merge it very soon. :-)

**Comment by StephanTLavavej** [comment]
For the third time today, thank you for the continual `span` conformance updates! 😸😺😸

**Comment by miscco** [comment]
It kind of is my baby so thanks a lot for the help and double checking.
What was kind of strange is that I actually build the test with the correct include:
```bash
MyInclude>set INCLUDE=MyInclude\STL\out\build\x64\out\inc;%INCLUDE%
MyInclude>set LIB=MyInclude\STL\out\build\x64\out\lib\amd64;%LIB%
MyInclude>set PATH=MyInclude\STL\out\build\x64\out\bin\amd64;%PATH%

C:\Users\bion\Desktop>cl /nologo /EHsc /W4 /WX /MDd /std:c++latest \I MyInclude\STL\tests\std\include MyInclude\STL\tests\std\tests\P0122R7_span\test.cpp
```

This build for me even with the changes in the span.h file and only after a complete rebuild of the STL did the error occur. 

The wonders of building a c++ TU

**Comment by StephanTLavavej** [comment]
You need `/I`, not `\I`, if that was your verbatim command line. And even with that fix, you’re adding `tests/std/include` (which few tests need), not overlaying `stl/inc` which is what’s necessary to consume header changes without rebuilding. Hope this helps!

**Pull request merged**